### ISSUE TITLE
Copy variable for every step leads to slow training speed?

### ISSUE THREAD

**Issue by PWZER** [keras-team/keras#15376]
**Describe the problem**.

I create a model with keras, but the training was too slow, and the memory usage is always changing. **ReadVariableOp** running time is too long.
After debugging, I found that the following line of code is executed every step.

https://github.com/tensorflow/tensorflow/blob/206301031d0051152a104da93fc637f2518881e1/tensorflow/core/kernels/resource_variable_ops.cc#L158

**Standalone code to reproduce the issue**.

[Colab Reproduce Code](https://colab.research.google.com/drive/1FMQ3dKPzXRAYOKi_n64hRxGHXAScXPZa?usp=sharing)

**Source code / logs**.

![timeline](https://user-images.githubusercontent.com/13550831/133596322-4cede011-ba38-4f29-ab1e-1c46b8a73964.png)

**System Information**
- Python Version: 3.8
- TensorFlow version: v2.5.0-160-g8222c1cfc86 2.5.1
- OS: ubuntu 20.04

**Comment by PWZER** [comment]
I think the copy-on-read is not suitable for larger variables, For larger variables, the performance of using shared locks will be better？

**Comment by jvishnuvardhan** [comment]
I was able to reproduce the issue. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/63f9ed182854ad7732ca265cca5c0d2a/test.ipynb).

output from the code is as shown below

### Training with Estimator

```
INFO:tensorflow:loss = 0.027672563, step = 50 (0.357 sec)
INFO:tensorflow:global_step/sec: 25.7492
INFO:tensorflow:loss = 0.025328398, step = 60 (0.389 sec)
INFO:tensorflow:global_step/sec: 24.8684
INFO:tensorflow:loss = 0.023781756, step = 70 (0.402 sec)
INFO:tensorflow:global_step/sec: 24.6785
INFO:tensorflow:loss = 0.022232072, step = 80 (0.408 sec)
INFO:tensorflow:global_step/sec: 26.0616
INFO:tensorflow:loss = 0.021159796, step = 90 (0.382 sec)
```

### Trainning with Keras
`100/100 [==============================] - 81s 697ms/step - loss: 0.8490`

Analysis (As mentioned by @PWZER )
We can see the Keras trainning speed is 697ms/step, but the Estimator trainning speed is 38ms/step, About 20 times.

**Comment by tilakrayal** [comment]
@PWZER,
I tried to execute the given code on tensorflow nightly version and found that Keras training speed is 37ms/step. It is similar to the Estimator training speed. Kindly find the gist of it [here](https://colab.research.google.com/gist/jvishnuvardhan/63f9ed182854ad7732ca265cca5c0d2a/test.ipynb) and confirm if the issue still persists. Thank you!

**Comment by PWZER** [comment]
> @PWZER, I tried to execute the given code on tensorflow nightly version and found that Keras training speed is 37ms/step. It is similar to the Estimator training speed. Kindly find the gist of it [here](https://colab.research.google.com/gist/jvishnuvardhan/63f9ed182854ad7732ca265cca5c0d2a/test.ipynb) and confirm if the issue still persists. Thank you!

@tilakrayal  Thank you very much for your reply!

Then I compared the two versions of tf2.5.1 and tf-nightly again, the test results are as follows

## tf2.5.1(v2.5.0-160-g8222c1cfc86)

- Training with Estimator

  ```bash
  INFO:tensorflow:loss = 0.029063284, step = 40 (0.279 sec)
  INFO:tensorflow:global_step/sec: 27.5201
  INFO:tensorflow:loss = 0.0259695, step = 50 (0.359 sec)
  INFO:tensorflow:global_step/sec: 26.4991
  INFO:tensorflow:loss = 0.023996644, step = 60 (0.378 sec)
  INFO:tensorflow:global_step/sec: 29.5931
  INFO:tensorflow:loss = 0.02239011, step = 70 (0.338 sec)
  INFO:tensorflow:global_step/sec: 29.1892
  INFO:tensorflow:loss = 0.0210438, step = 80 (0.342 sec)
  INFO:tensorflow:global_step/sec: 27.8276
  INFO:tensorflow:loss = 0.020100228, step = 90 (0.359 sec)
  ```

- Trainning with Keras

```bash
100/100 [==============================] - 31s 226ms/step - loss: 0.0619
```

## tf-nightly(v1.12.1-83306-g36495debc6c)

- Training with Estimator
  
  ```bash
  INFO:tensorflow:loss = 0.40406987, step = 40 (2.514 sec)
  INFO:tensorflow:global_step/sec: 4.0039
  INFO:tensorflow:loss = 0.40406987, step = 50 (2.499 sec)
  INFO:tensorflow:global_step/sec: 4.01341
  INFO:tensorflow:loss = 0.40406987, step = 60 (2.494 sec)
  INFO:tensorflow:global_step/sec: 3.56581
  INFO:tensorflow:loss = 0.40406987, step = 70 (2.802 sec)
  INFO:tensorflow:global_step/sec: 3.98933
  INFO:tensorflow:loss = 0.40406987, step = 80 (2.506 sec)
  INFO:tensorflow:global_step/sec: 3.97352
  INFO:tensorflow:loss = 0.40406987, step = 90 (2.516 sec)
  ```
- Trainning with Keras

  ```bash
  100/100 [==============================] - 32s 250ms/step - loss: 0.1840
  ```

From the results, the speed of keras has been improved in the tf-nightly version, which is the same as the tf2.5.1 version. But the Estimator is much slower, why is this?

**Comment by tilakrayal** [comment]
@PWZER,
We can see that the tf-nightly version you are using v1.12.1, whereas if you check the gist attached, on tf-nightly version Keras training speed is 37ms/step. It is similar to the Estimator training speed. Kindly find the gist of it [here](https://colab.research.google.com/gist/jvishnuvardhan/63f9ed182854ad7732ca265cca5c0d2a/test.ipynb). Thank you!

**Comment by PWZER** [comment]
> @PWZER, We can see that the tf-nightly version you are using v1.12.1, whereas if you check the gist attached, on tf-nightly version Keras training speed is 37ms/step. It is similar to the Estimator training speed. Kindly find the gist of it [here](https://colab.research.google.com/gist/jvishnuvardhan/63f9ed182854ad7732ca265cca5c0d2a/test.ipynb). Thank you!

Yes, I just run the gist script you gave, but i don't see the result you said. you can look at the output [here](https://colab.research.google.com/gist/PWZER/7b779a2c0673a6422832c7a8ac2f754c/test.ipynb).

**Comment by tilakrayal** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new `Keras 3` code base from **keras-team/keras-core** to **keras-team/keras.**
Consequently, This issue may not be relevant to the `Keras 3 `code base. After the migration is successfully completed, feel free to reopen this issue at **keras-team/keras** if you believe it remains relevant to the `Keras 3` code base. 
If instead this issue is a bug or security issue in legacy **tf.keras**, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about `Keras 3,` please take a look at https://keras.io/keras_core/announcement/. Thank you!
